### PR TITLE
ETH price hook

### PR DESCRIPTION
**Pull request by codenamejason**
Tasks Completed:
- Add `useEthPrice` hook file
- Basic call to Uniswap TWAP
- Working example shows in UI balance

### CODE REVIEW

**Comment by codenamejason** [comment]
@carletex I would also like to write a test file for this hook, how should we structure the tests?

Ex:
`/packages/frontend/__tests__/useEthHook.test.ts`

And would Jest be an option or did you have something in mind?

File: packages/frontend/components/hooks/useEthPrice.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Fetcher, Pair, Route, Token, WETH } from "@uniswap/sdk";
+
+export const useEthPrice = async (provider: any): Promise<number> => {
+  // Get the price of ETH based on ETH/DAI trading pair from Uniswap SDK
```
Comment (carletex) [inline packages/frontend/components/hooks/useEthPrice.ts:4]: You can move this comment above the function declation.

Check others hooks/component to copy the "comment style"

File: packages/frontend/pages/index.tsx [hunk 2]
```diff
@@ -13,9 +16,23 @@ const tempAutoConnectConfig: TAutoConnect = {
 };
 
 const Home: NextPage = () => {
+  const [price, setPrice] = useState<number>(0);
+  const mainnetProvider = useProvider({ chainId: 1 });
```
Comment (carletex) [inline packages/frontend/pages/index.tsx:20]: This should be inside the hook. We always want to get the ETH price from mainnet.

**Review by carletex — COMMENTED** [review]
Hey @codenamejason Thanks for this.

Left some review comments.

Regarding the tests: Yep, it's something we'd want to add. I was thinking to make them when the MVP is ready. But feel free to create an issue and work on it in a different PR. They can live under `frontend/test`... and yeah, Jest should be fine for now.

File: packages/frontend/pages/index.tsx [hunk 3]
```diff
@@ -62,7 +80,7 @@ const Home: NextPage = () => {
         </p>
 
         <h3 className="font-bold">Balance Component</h3>
-        <Balance address="0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266" price={1350} />
+        <Balance address="0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266" price={price} />
```
[the remark this answers is not in the document]
Comment (codenamejason) [reply packages/frontend/pages/index.tsx:83]: Right so we will wanna dispatch a state update when the eth price is fetched in the Hook?

File: packages/frontend/hooks/scaffold-eth/useEthPrice.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import { Fetcher, Route, Token, WETH } from "@uniswap/sdk";
+import { useEffect, useState } from "react";
+import { useProvider } from "wagmi";
+
+// Get the price of ETH based on ETH/DAI trading pair from Uniswap SDK
```
Comment (carletex) [inline packages/frontend/hooks/scaffold-eth/useEthPrice.ts:5]: Check how we put comments in functions (JSDoc style). We want to keep it consistent.

```js
/**
 * Get the price of ETH based on ETH/DAI trading pair from Uniswap SDK
 */
```

File: packages/frontend/hooks/scaffold-eth/useEthPrice.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+export const useEthPrice = () => {
+  const provider = useProvider({ chainId: 1 });
+  const [ethPrice, setEthPrice] = useState<number>(0);
```
Comment (carletex) [inline packages/frontend/hooks/scaffold-eth/useEthPrice.ts:8]: No need to specify the type. It's inferred from the default value.

File: packages/frontend/hooks/scaffold-eth/useEthPrice.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+  useEffect(() => {
+    Promise.resolve(
+      fetchPriceFromUniswap(provider).then((price: any) => {
+        console.log("price", price);
+        setEthPrice(price);
+      }),
+    );
+  }, [provider]);
```
Comment (carletex) [inline packages/frontend/hooks/scaffold-eth/useEthPrice.ts:17]: We could do everything in the useEffect.

- No need to pass the provider as an argument, we already have it in the state.

```ts
  useEffect(() => {
    const fetchPriceFromUniswap = async () => {
      try {
        const DAI = new Token(1, "0x6B175474E89094C44Da98b954EedeAC495271d0F", 18);
        const pair = await Fetcher.fetchPairData(DAI, WETH[DAI.chainId], provider);
        const route = new Route([pair], WETH[DAI.chainId]);
        const price = parseFloat(route.midPrice.toSignificant(6));

        setEthPrice(price);
      } catch (error) {
        console.log("Error fetching ETH price from Uniswap: ", error);
      }
    };
    fetchPriceFromUniswap(); // no need to await here.
  }, [provider]);
```

File: packages/hardhat/.gitignore [hunk 7]
```diff
@@ -11,3 +11,4 @@ artifacts
 
 deployments
 
+/*.txt
```
Comment (carletex) [inline packages/hardhat/.gitignore:14]: I think this doesn't belong to this PR

**Review by carletex — CHANGES_REQUESTED** [review]
Thanks Jason! Left some comments

File: packages/hardhat/.gitignore [hunk 8]
```diff
@@ -10,4 +10,3 @@ cache
 artifacts
 
 deployments
-
```
Comment (carletex) [inline packages/hardhat/.gitignore:13]: You might want to leave this as it was, so there is no changes to this file

File: packages/frontend/hooks/scaffold-eth/useEthPrice.ts (hunk 6 above)
Comment (carletex) [reply packages/frontend/hooks/scaffold-eth/useEthPrice.ts]: Unresolving this. I don't see this changes implemented. Maybe you forgot to push?

File: packages/frontend/pages/index.tsx [hunk 9]
```diff
@@ -2,6 +2,7 @@ import { ConnectButton } from "@rainbow-me/rainbowkit";
 import type { NextPage } from "next";
```
Comment (carletex) [inline packages/frontend/pages/index.tsx:2]: There are some prettier issues in this file, so the lint tests are failing.

image

**Review by carletex — CHANGES_REQUESTED** [review]
Left some comments & "unresolved" a previous comment that haven't been fixed.

File: packages/frontend/hooks/scaffold-eth/useEthPrice.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  const [ethPrice, setEthPrice] = useState(0);
+
+  useEffect(() => {
+    Promise.resolve(
+      fetchPriceFromUniswap(provider).then((price: any) => {
+        console.log("price", price);
+        setEthPrice(price);
+      }),
+    );
+  }, []);
```
Comment (codenamejason) [inline packages/frontend/hooks/scaffold-eth/useEthPrice.ts:20]: it's been removed @carletex

File: packages/hardhat/.gitignore (hunk 8 above)
Comment (codenamejason) [reply packages/hardhat/.gitignore:13]: this is how it was.. not sure how to remove the history

File: packages/frontend/hooks/scaffold-eth/useEthPrice.ts (hunk 6 above)
Comment (codenamejason) [reply packages/frontend/hooks/scaffold-eth/useEthPrice.ts]: this is not what i see

File: packages/hardhat/.gitignore (hunk 8 above)
Comment (carletex) [reply packages/hardhat/.gitignore:13]: Just add a new line in the file (as it was before) so It won't show on the diff.

image

File: packages/frontend/hooks/scaffold-eth/useEthPrice.ts (hunk 6 above)
Comment (carletex) [reply packages/frontend/hooks/scaffold-eth/useEthPrice.ts]: Not sure where you are looking, but this is what you are trying to merge ("files changed" tab)

image

Comment (codenamejason) [reply packages/frontend/hooks/scaffold-eth/useEthPrice.ts]: Now I'm confused lol I thought we wanted to remove provider from dependency array?

Comment (carletex) [reply packages/frontend/hooks/scaffold-eth/useEthPrice.ts]: Not from the dep array. You were calling `fetchPriceFromUniswap` with `provider` as an argument. And we don't need it to pass it over, since it's already in the state.

I think the code I pasted above should do it.

Comment (codenamejason) [reply packages/frontend/hooks/scaffold-eth/useEthPrice.ts]: Gotcha 😝

File: packages/hardhat/.gitignore (hunk 8 above)
Comment (codenamejason) [reply packages/hardhat/.gitignore:13]: So strange... I have the extra line in my editor and shows no changes.

Screen Shot 2022-10-22 at 12 32 53 PM

Comment (carletex) [reply packages/hardhat/.gitignore:13]: Maybe your editor is configured to strip the last line when saving?

Anyway, no worries, I can update that. I'll review & test again on Monday.

Thanks Jason!

**Review by carletex — APPROVED** [review]
Made some minor tweaks https://github.com/scaffold-eth/se-2/pull/22/commits/1bde56167a6b3b1ef54de2fd205b4880aacc7bd9

Ready to merge! thanks :)
